machine_learning/k_means_clustering.c

```diff
@@ -0,0 +1,362 @@
+{
+    cluster* clusters = NULL;
+    if (k <= 1)
+    {
+        /*
+        If we have to cluster them only in one group
+        then calculate centroid of observations and
+        that will be a ingle cluster
+        */
+        clusters = (cluster*)malloc(sizeof(cluster));
+        memset(clusters, 0, sizeof(cluster));
+        calculateCentroid(observations, size, clusters);
+    }
+    else if (k < size)
+    {
+        clusters = malloc(sizeof(cluster) * k);
+        memset(clusters, 0, k * sizeof(cluster));
+        /* STEP 1 */
+        for (size_t j = 0; j < size; j++)
+        {
+            observations[j].group = rand() % k;
+        }
+        size_t changed = 0;
+        size_t minAcceptedError =
+            size /
+            10000;  // Do until 99.99 percent points are in correct cluster
+        int t = 0;
+        do
+        {
+            /* Initialize clusters */
+            for (int i = 0; i < k; i++)
+            {
+                clusters[i].x = 0;
+                clusters[i].y = 0;
+                clusters[i].count = 0;
+            }
+            /* STEP 2*/
+            for (size_t j = 0; j < size; j++)
+            {
+                t = observations[j].group;
+                clusters[t].x += observations[j].x;
+                clusters[t].y += observations[j].y;
+                clusters[t].count++;
+            }
+            for (int i = 0; i < k; i++)
+            {
+                clusters[i].x /= clusters[i].count;
+                clusters[i].y /= clusters[i].count;
+            }
+            /* STEP 3 and 4 */
+            changed = 0;  // this variable stores change in clustering
+            for (size_t j = 0; j < size; j++)
+            {
+                t = calculateNearst(observations + j, clusters, k);
+                if (t != observations[j].group)
+                {
+                    changed++;
+                    observations[j].group = t;
+                }
+            }
+        } while (changed > minAcceptedError);  // Keep on grouping until we have
+                                               // got almost best clustering
+    }
+    else
+    {
+        /* If no of clusters is more than observations
+           each observation can be its own cluster
+        */
+        clusters = (cluster*)malloc(sizeof(cluster) * k);
+        memset(clusters, 0, k * sizeof(cluster));
+        for (int j = 0; j < size; j++)
+        {
+            clusters[j].x = observations[j].x;
+            clusters[j].y = observations[j].y;
+            clusters[j].count = 1;
+            observations[j].group = j;
+        }
+    }
+    return clusters;
+}
+
+/*! @fn test
+ * A function to test the kMeans function
+ * Generates 100000 points in a circle of
+ * radius 20.0 with center at (0,0)
+ * and cluster them into 5 clusters
+ *
+ * ![Sample
+ * Output](https://raw.githubusercontent.com/TheAlgorithms/C/docs/images/machine_learning/kMeansTest1.png)
+ */
+void test()
```

Comment:
```suggestion
static void test()
```
To avoid the CI tools to think that this function has global visibility